### PR TITLE
Feature/as 5882 lpa account enter code

### PR DESCRIPTION
**Pull request by suityou01**
## Ticket Number

https://pins-ds.atlassian.net/browse/AS-5882

## Description of change

Add flow for LPA user to enter the code sent to them by email

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/controllers/common/enter-code.js [hunk 1]
```diff
@@ -131,7 +201,97 @@ const postEnterCode = (views) => {
...
+				logger.error(e);
+				return {
+					valid: false,
+					action: enterCodeConfig.actions.lpaDashboard
+				};
+			}
+		}
+
+		// check token
+		let token = await isTokenValid(id, emailCode, req.session, id);
```
Comment (bryn-pins) [inline packages/forms-web-app/src/controllers/common/enter-code.js:275]: Don't think the 4th param is used here, unless I've missed something

File: packages/forms-web-app/src/lib/appeals-api-wrapper.js [hunk 2]
```diff
@@ -149,6 +149,15 @@ exports.createUser = async (email, isAdmin, lpaCode) => {
 	});
 };
 
+exports.getUserById = async (id) => {
+	console.log(`/api/v1/users/${id}`);
```
Comment (bryn-pins) [inline packages/forms-web-app/src/lib/appeals-api-wrapper.js:153]: can remove this console log?

File: packages/forms-web-app/src/lib/is-token-valid.js [hunk 3]
```diff
@@ -1,45 +1,54 @@
...
-	if (!id || typeof id !== 'string' || !token || typeof token !== 'string') {
-		return result;
-	}
-
-	// is test LPA + Test environment + test token
-	// then allow through as a confirm email token
-	if (
+// is test LPA + Test environment + test token
+// then allow through as a confirm email token
+const isTestLPAAndEnvironment = (token, session) => {
```
Comment (bryn-pins) [inline packages/forms-web-app/src/lib/is-token-valid.js:10]: think this should be renamed to include that it is checking the token, or perhaps a suggestion to be separated into 2 functions if name is too unwieldy

File: packages/forms-web-app/src/validators/common/check-id-is-email-address.js
Comment (bryn-pins) [inline packages/forms-web-app/src/validators/common/check-id-is-email-address.js:1]: is this being used?

File: packages/forms-web-app/src/validators/lpa/enter-code.js
Comment (bryn-pins) [inline packages/forms-web-app/src/validators/lpa/enter-code.js:1]: More of a question, this is taking into account the changes to the code in the other PR

File: packages/forms-web-app/src/controllers/common/enter-code.js (hunk 1 above)
Comment (suityou01) [reply packages/forms-web-app/src/controllers/common/enter-code.js:275]: Or the 3rd. I wonder if it read req.session.id at some point... Good spot.

File: packages/forms-web-app/src/validators/lpa/enter-code.js
Comment (bryn-pins) [reply packages/forms-web-app/src/validators/lpa/enter-code.js:1]: Sorry not clear: https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/2894

Comment (suityou01) [reply packages/forms-web-app/src/validators/lpa/enter-code.js:1]: I have not integration tested it, but as the variable that is storing the code value is a string I don't see any issue. What did I miss?

Comment (suityou01) [reply packages/forms-web-app/src/validators/lpa/enter-code.js:1]: Phew! :-)

**Comment by bryn-pins** [comment]
Think this looks good, only concern is we're not following the same pattern as the other enter code pages of sending the token to the user's email when hitting the GET for enter-code

Think it is perhaps more logical not to, but then will need to ensure the token is sent before we get to this page, including on the various, need-new-code, code expired etc. routes

File: packages/forms-web-app/src/routes/lpa-dashboard/lpa-dashboard.js [hunk 4]
```diff
@@ -7,6 +7,7 @@ const { getServiceInvite } = require('../../../src/controllers/lpa-dashboard/ser
 const router = express.Router();
 
 router.use(featureFlagMiddleware('lpa-dashboard'));
+
 router.use(
 	skipMiddlewareForPaths(requireUser, [
 		'input-code',
```
Comment (chrisHowarthkc) [inline packages/forms-web-app/src/routes/lpa-dashboard/lpa-dashboard.js:13]: Should 'input-code' be replaced with 'enter-code' for this journey?

**Comment by bryn-pins** [comment]
Ok so if we're keeping the same as the other enter token pages, I think we need to actually generate the token on the get for enter-code

So a call to: sendToken(userId, action, userEmail)

In order to get userEmail, we can lookup via the getUserById api method for now

**Comment by bryn-pins** [comment]
Happy to approve given outside conversation once conflicts are resolved

**Review by bryn-pins — APPROVED** [review]

File: packages/forms-web-app/src/lib/is-token-valid.js [hunk 5]
```diff
@@ -1,45 +1,52 @@
...
 		} else {
 			throw err;
 		}
 	}
+};
+
+const isTokenValid = async (id, token) => {
+	let result = {
+		valid: false,
+		action: ''
```
[the remark this answers is not in the document]
Comment (chrishawldev) [reply packages/forms-web-app/src/lib/is-token-valid.js:97]: Cool, I can see this is handled here const getToken = async (id, token, session) https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/2908/files#diff-732692fed0da9475bd6db5d66bc33777d85ec9b06a11e34f2bfa883cff722e2bR29
if (err.message === 'Too Many Requests') {
			return {
				valid: false,
				action: session?.enterCode?.action,
				tooManyAttempts: true
			};

**Review by chrisHowarthkc — APPROVED** [review]

**Pull request merged**